linear_operator/operators/permutation_linear_operator.py

```diff
@@ -0,0 +1,133 @@
+from math import isqrt
+from typing import Optional
+
+import torch
+from torch import Tensor
+
+from ._linear_operator import LinearOperator
+
+
+class AbstractPermutationLinearOperator(LinearOperator):
+    r"""Abstract base class for permutation operators.
+    Incorporates 1) square shape, 2) common input shape checking, and
+    3) the fact that permutation matrices' transposes are their inverses.
+    """
+
+    def inverse(self):
+        return self._transpose_nonbatch()
+
+    def _solve(self, rhs: Tensor):
+        self._check_shape(rhs)
+        return self.inverse() @ rhs
+
+    def _size(self):
+        n = len(self)
+        return (n, n)
+
+    def _check_shape(self, rhs) -> None:
+        if rhs.shape[-2] != len(self):
+            raise ValueError(
+                f"rhs.shape[-2] = {rhs.shape[-2]} incompatible " f"with size of permutation operator {len(self)}."
+            )
+
+
+class PermutationLinearOperator(AbstractPermutationLinearOperator):
+    r"""LinearOperator that lazily represents a permutation matrix with O(n) memory.
+
+    Args:
+        - perm: A permutation tensor with which to permute the first non-batch
+            dimension through matmul. Should have integer elements.
+        - inv_perm: Optional tensor representing the inverse of perm, is computed
+            via a O(n log(n)) sort if not given.
+        - check: Boolean
+    """
+
+    def __init__(
+        self,
+        perm: Tensor,
+        inv_perm: Optional[Tensor] = None,
+        check: bool = True,
+    ):
+        if not isinstance(perm, Tensor):
+            raise ValueError("perm is not a Tensor.")
+
+        if not perm.ndim == 1:
+            raise ValueError("perm is not a 1d Tensor.")
+
+        if inv_perm is not None:
+            sorted_perm = perm[inv_perm]
+        else:
+            sorted_perm, inv_perm = perm.sort()
+
+        if check:
+            for i, si in enumerate(sorted_perm):
+                if torch.is_floating_point(si) or torch.is_complex(si):
+                    raise ValueError("perm does not have integer elements.")
+                if sorted_perm[i] != i:
```

Comment:
If `inv_perm` is `None`, `sorted_perm` is guaranteed to be ordered correctly, but this can still fail if `perm` is not actually a permutation. I am testing this case with 
```python
with self.assertRaisesRegex(ValueError, "Invalid perm*"):
     PermutationLinearOperator(torch.tensor([1, 3, 5]))
```
in the test file. So I think there is still value in keeping this test. What do you think?